src/IO/H5/Cce.cpp

```diff
@@ -0,0 +1,262 @@
+  }
+
+  if (exists) {
+    {
+      // We treat this as an internal version for now. We'll need to deal with
+      // proper versioning later.
+
+      // Check if the version exists before calling the open_version.
+      const htri_t version_exists = H5Aexists(cce_group_.id(), "version.ver");
+      if (version_exists != 0) {
+        const Version open_version(true, detail::OpenGroup{}, cce_group_.id(),
+                                   "version");
+        version_ = open_version.get_version();
+      }
+    }
+    {
+      const htri_t header_exists = H5Aexists(cce_group_.id(), "header.hdr");
+      if (header_exists != 0) {
+        const Header header(true, detail::OpenGroup{}, cce_group_.id(),
+                            "header");
+        header_ = header.get_header();
+      }
+    }
+
+    for (const std::string& bondi_var : bondi_variables_) {
+      DataSet& dataset = bondi_datasets_.at(bondi_var);
+      dataset.id =
+          H5Dopen2(cce_group_.id(), bondi_var.c_str(), h5::h5p_default());
+      CHECK_H5(dataset.id, "Failed to open dataset");
+
+      hid_t space_id = H5Dget_space(dataset.id);
+      std::array<hsize_t, 2> max_dims{};
+      if (2 != H5Sget_simple_extent_dims(space_id, dataset.size.data(),
+                                         max_dims.data())) {
+        ERROR("Invalid number of dimensions in cce file " << bondi_var
+                                                          << " on disk.");
+      }
+      CHECK_H5(H5Sclose(space_id), "Failed to close dataspace");
+
+      if (legend_ != read_rank1_attribute<std::string>(dataset.id, "Legend"s)) {
+        ERROR("l_max from cce file " << bondi_var
+                                     << " does not match l_max in constructor");
+      }
+    }
+  } else {  // file does not exist
+    {
+      Version open_version(false, detail::OpenGroup{}, cce_group_.id(),
+                           "version", version_);
+    }
+    {
+      Header header(false, detail::OpenGroup{}, cce_group_.id(), "header");
+      header_ = header.get_header();
+    }
+    // So this file is compatible with the sxs/scri python packages offered by
+    // the SXS collaboration
+    write_to_attribute(cce_group_.id(), sxs_format_str_, sxs_version_str_);
+
+    for (const std::string& bondi_var : bondi_variables_) {
+      DataSet& dataset = bondi_datasets_.at(bondi_var);
+      dataset.id = h5::detail::create_extensible_dataset(
+          cce_group_.id(), bondi_var, dataset.size,
+          std::array<hsize_t, 2>{{4, legend_.size()}},
+          {{h5s_unlimited(), legend_.size()}});
+      CHECK_H5(dataset.id, "Failed to create dataset");
+
+      write_to_attribute(dataset.id, "Legend"s, legend_);
+      // So this dataset is compatible with the sxs/scri python packages offered
+      // by the SXS collaboration
+      write_to_attribute(dataset.id, sxs_format_str_, sxs_version_str_);
+    }
+  }
+}
+
+Cce::~Cce() {
+#ifdef __clang__
```

Comment:
Why are we avoiding structured bindings with clang and using them with gcc?  If the point of this `#ifdef` is just to protect the pragmas, just put it around them and use the same normal c++ for both versions, otherwise comment what's going on.

Reply:
My compiler (clang 14) complained about trying to capture structured bindings in a lambda (in the ERROR macro I believe, which CHECK_H5 calls). GCC didn't care about this. The `#ifdef` was more for the pragmas. I can just avoid the structured bindings, add a comment explaining why, and only add the `#ifdef` for the pragmas.

Reply:
I think there is a C++17-ish bug about this that gets fixed in C++20 (or 23), so whether or not we can directly capture struct bindings in lambdas currently depends on the compiler :(